python/lsst/drp/tasks/wcsfit.py

```diff
@@ -0,0 +1,1019 @@
+        inputDict = {}
+        modelComponents = ["INSTRUMENT/DEVICE", "EXPOSURE"]
+        baseMap = {"Type": "Composite", "Elements": modelComponents}
+        inputDict["EXPOSURE/DEVICE/base"] = baseMap
+
+        xMin = str(inputVisitSummary['bbox_min_x'].min())
+        xMax = str(inputVisitSummary['bbox_max_x'].max())
+        yMin = str(inputVisitSummary['bbox_min_y'].min())
+        yMax = str(inputVisitSummary['bbox_max_y'].max())
+
+        deviceModel = {"Type": "Composite", "Elements": self.config.deviceModel.list()}
+        inputDict["INSTRUMENT/DEVICE"] = deviceModel
+        for component in self.config.deviceModel:
+            if "poly" in component.lower():
+                componentDict = {"Type": "Poly",
+                                 "XPoly": {"OrderX": self.config.devicePolyOrder,
+                                           "SumOrder": True},
+                                 "YPoly": {"OrderX": self.config.devicePolyOrder,
+                                           "SumOrder": True},
+                                 "XMin": xMin, "XMax": xMax, "YMin": yMin, "YMax": yMax}
+            elif "identity" in component.lower():
+                componentDict = {"Type": "Identity"}
+            else:
+                raise ValueError(f"deviceModel component {component} is not supported")
+            inputDict[component] = componentDict
+
+        exposureModel = {"Type": "Composite", "Elements": self.config.exposureModel.list()}
+        inputDict["EXPOSURE"] = exposureModel
+        for component in self.config.exposureModel:
+            if "poly" in component.lower():
+                componentDict = {"Type": "Poly",
+                                 "XPoly": {"OrderX": self.config.exposurePolyOrder,
+                                           "SumOrder": "true"},
+                                 "YPoly": {"OrderX": self.config.exposurePolyOrder,
+                                           "SumOrder": "true"}}
+            elif "identity" in component.lower():
+                componentDict = {"Type": "Identity"}
+            else:
+                raise ValueError(f"exposureModel component {component} is not supported")
+            inputDict[component] = componentDict
+
+        inputYAML.addInput(yaml.dump(inputDict))
+        inputYAML.addInput("Identity:\n  Type:  Identity\n")
+
+        return inputYAML
+
+    def _addObjects(self, wcsf, inputCatalogRefs):
+        """Add science sources to the wcsfit.WCSFit object.
+
+        Parameters
+        ----------
+        wcsf : `wcsfit.WCSFit`
+            WCS-fitting object
+        inputCatalogRefs : `list`
+            List of DeferredDatasetHandles pointing to visit-level source
+            tables
+        """
+        for inputCatalogRef in inputCatalogRefs:
+            visit = inputCatalogRef.dataId['visit']
+
+            inputCatalog = inputCatalogRef.get(parameters={'columns': self.columns})
```

Comment:
It looks like you're reading these catalogs twice, once here and once in `_load_catalogs_and_associate`. Is that an intentional tradeoff between memory and disk reads? I'd be rather surprised that the catalogs themselves are a significant part of the memory footprint: for jointcal, the full fit matrix was vastly larger, and the catalogs basically were irrelevant in comparison.

Reply:
Yeah, I don't want to keep the catalogs in memory, and the disk read time is relatively short (about 0.05 seconds per visit). I haven't tested how it affects performance, but it definitely would be a significant part of the memory footprint. Remember that a big difference between `jointcal` and `gbdes` is that the star positions are not fit parameters in `gbdes`, so the fit matrix is actually not that big.